### PR TITLE
[IRGenSIL] When emitting a shadow copy, honour the layout.

### PR DESCRIPTION
**Pull request by dcci**
This code was ad-hoc trying to reconstruct the layout. Ask TypeInfo
instead, as it knows what's the right thing to do.
This allows to print field of classes nested inside classes correctly
in lldb (SR-6791).

<rdar://problem/36518505>

### CODE REVIEW

**Comment by dcci** [comment]
Bonus, this makes the code easier to read/shorter.

**Comment by dcci** [comment]
@swift-ci test and merge

File: test/DebugInfo/guard-let.swift [hunk 1]
```diff
@@ -8,12 +8,9 @@ func use<T>(_ t: T) {}
...
 {
   // CHECK: define {{.*}}@"$S4main1fyySiSgF"
-  // The shadow copy store should not have a location.
-  // Note that the store must be in the same scope or else it might defeat
-  // livedebugvalues.
+  // CHECK1: %debug.copy = alloca %TSiSg
+  // CHECK1: @llvm.dbg.declare(metadata %TSiSg* %debug.copy
   // CHECK1: @llvm.dbg.declare(metadata {{(i32|i64)}}* %val.addr, {{.*}}, !dbg ![[DBG0:.*]]
-  // CHECK1: %[[PHI:.*]] = phi
-  // CHECK1: store {{(i32|i64)}} %[[PHI]], {{(i32|i64)}}* %val.addr, align {{(4|8)}}, !dbg ![[DBG1:.*]]
```
Comment (vedantk) [inline test/DebugInfo/guard-let.swift]: Shouldn't we still store a shadow value for the phi somewhere? Can we keep checking for that?

**Review by vedantk — CHANGES_REQUESTED** [review]

**Comment by dcci** [comment]
For reference, this is the generated IR:
```
define swiftcc void @"$S4main1fyySiSgF"(i64, i8) #2 !dbg !48 {
entry:
  %debug.copy = alloca %TSiSg, align 8
  call void @llvm.dbg.declare(metadata %TSiSg* %debug.copy, metadata !52, metadata !DIExpression()),
!dbg !53
  %val.addr = alloca i64, align 8
  call void @llvm.dbg.declare(metadata i64* %val.addr, metadata !54, metadata !DIExpression()), !dbg
!57
  %2 = alloca %TSi, align 8
  %3 = trunc i8 %1 to i1
  %4 = bitcast %TSiSg* %debug.copy to i8*, !dbg !58
  call void @llvm.lifetime.start.p0i8(i64 9, i8* %4), !dbg !58
  %5 = bitcast %TSiSg* %debug.copy to i64*, !dbg !58
  store i64 %0, i64* %5, align 8, !dbg !58
  %6 = getelementptr inbounds %TSiSg, %TSiSg* %debug.copy, i32 0, i32 1, !dbg !58
  %7 = bitcast [1 x i8]* %6 to i1*, !dbg !58
  store i1 %3, i1* %7, align 8, !dbg !58
  br i1 %3, label %9, label %8, !dbg !59

; <label>:8:                                      ; preds = %entry
  br label %10, !dbg !59

; <label>:9:                                      ; preds = %entry
  br label %16, !dbg !59

; <label>:10:                                     ; preds = %8
  %11 = phi i64 [ %0, %8 ], !dbg !60
  store i64 %11, i64* %val.addr, align 8, !dbg !60
  br label %12, !dbg !59
```

**Comment by dcci** [comment]
I and Vedant discussed this offline.
We agreed that this code used to check that we store a value to the `alloca` we create, and now the test doesn't anymore. I'll add the correct check lines, i.e.

```
 // CHECK: %5 = bitcast %TSiSg* %debug.copy to i64*, !dbg !58
 // CHECK: store i64 %0, i64* %5, align 8, !dbg !58
```

**Pull request merged**